File: stoichiometry.R
library("stoichcalc")


# --- simple stoichiometric solution --- #

HPO4 <- c(P = 1.0,   # gP/gHPO4-P
          N = 0.0)   # gN/gHPO4-P
NH4  <- c(P = 0.0,   # gP/gN4-N
          N = 1.0)   # gN/gN4-N
ALG  <- c(P = 0.01,  # gP/gALG
          N = 0.1)   # gN/gALG

subst.comp <- list(HPO4 = HPO4,
                   NH4  = NH4,
                   ALG  = ALG)

# build composition matrix
alpha <- calc.comp.matrix(subst.comp)
print(alpha)

# calculate stoichiometric coefficients
nu.growth.ALG <- calc.stoich.coef(alpha      = alpha,
                                  name       = "growth.ALG",
                                  subst      = c("HPO4", "NH4", "ALG"),
                                  subst.norm = "ALG",
                                  nu.norm    = 1)
nu.resp.ALG <- calc.stoich.coef(alpha      = alpha,
                                name       = "resp.ALG",
                                subst      = c("HPO4", "NH4", "ALG"),
                                subst.norm = "ALG",
                                nu.norm    = -1)
nu <- rbind(nu.growth.ALG, nu.resp.ALG)
print(nu)




# --- complex stoichiometric solution --- #

param <- list(alpha.O.ALG = 0.50,   # gO/gALG
              alpha.H.ALG = 0.07,   # gH/gALG
              alpha.N.ALG = 0.06,   # gN/gALG
              alpha.P.ALG = 0.005,  # gP/gALG
              alpha.O.ZOO = 0.50,   # gO/gZOO
              alpha.H.ZOO = 0.07,   # gH/gZOO
              alpha.N.ZOO = 0.06,   # gN/gZOO
              alpha.P.ZOO = 0.01,   # gP/gZOO
              alpha.O.POM = 0.40,   # gO/gPOM
              alpha.H.POM = 0.07,   # gH/gPOM
              alpha.N.POM = 0.04,   # gN/gPOM
              alpha.P.POM = 0.007,  # gP/gPOM
              Y.ZOO       = 0.2,    # gZOO/gALG
              f.e         = 0.4)    # gPOM/gALG

# define carbon factor such that factors sum to 1
param$alpha.C.ALG = 1 - (
    param$alpha.O.ALG
    + param$alpha.H.ALG
    + param$alpha.N.ALG
    + param$alpha.P.ALG
)
param$alpha.C.ZOO = 1 - (
    param$alpha.O.ZOO
    + param$alpha.H.ZOO
    + param$alpha.N.ZOO
    + param$alpha.P.ZOO
)
param$alpha.C.POM = 1 - (
    param$alpha.O.POM
    + param$alpha.H.POM
    + param$alpha.N.POM
    + param$alpha.P.POM
)

NH4  <- c(N = 1,
          H = 4,
          charge = 3)
NO3  <- c(N = 1,
          O = 3,
          charge = -1)
HPO4 <- c(H = 1,
          P = 1,
          O = 4,
          charge = -2)
HCO3 <- c(H = 1,
          C = 1,
          O = 3,
          charge = -1)
O2   <- c(O = 2,
          charge = 0)
H    <- c(H = 1,
          charge = 1)
H2O  <- c(H = 2,
          O = 1,
          charge = 0)

ALG <- c(O = param$alpha.O.ALG / 16,
         H = param$alpha.H.ALG / 1,
         N = param$alpha.N.ALG / 14,
         P = param$alpha.P.ALG / 31,
         C = param$alpha.C.ALG / 12)
ZOO <- c(O = param$alpha.O.ZOO / 16,
         H = param$alpha.H.ZOO / 1,
         N = param$alpha.N.ZOO / 14,
         P = param$alpha.P.ZOO / 31,
         C = param$alpha.C.ZOO / 12)
POM <- c(O = param$alpha.O.POM / 16,
         H = param$alpha.H.POM / 1,
         N = param$alpha.N.POM / 14,
         P = param$alpha.P.POM / 31,
         C = param$alpha.C.POM / 12)

subst.comp <- list(NH4  = NH4,
                   NO3  = NO3,
                   HPO4 = HPO4,
                   HCO3 = HCO3,
                   O2   = O2,
                   H    = H,
                   H2O  = H2O,
                   ALG  = ALG,
                   ZOO  = ZOO,
                   POM  = POM)

# build composition matrix
alpha.complex <- calc.comp.matrix(subst.comp)
print(alpha.complex)

# define additional params
param$Y.ALG.death = min(1,
                        param$alpha.N.ALG / param$alpha.N.POM,
                        param$alpha.P.ALG / param$alpha.P.POM)
param$Y.ZOO.death = min(1,
                        param$alpha.N.ZOO / param$alpha.N.POM,
                        param$alpha.P.ZOO / param$alpha.P.POM)

# calculate stoichiometric coefficients
nu.growth.ALG.NH4.complex <- calc.stoich.coef(alpha      = alpha.complex,
                                              name       = "growth.ALG.NH4",
                                              subst      = c("NH4", "HPO4", "HCO3", "O2", "H", "H2O", "ALG"),
                                              subst.norm = "ALG",
                                              nu.norm    = 1.0)
nu.growth.ALG.NO3.complex <- calc.stoich.coef(alpha      = alpha.complex,
                                              name       = "growth.ALG.NO3",
                                              subst      = c("NO3", "HPO4", "HCO3", "O2", "H", "H2O", "ALG"),
                                              subst.norm = "ALG",
                                              nu.norm    = 1.0)
nu.resp.ALG.complex <- calc.stoich.coef(alpha      = alpha.complex,
                                        name       = "resp.ALG",
                                        subst      = c("NH4", "HPO4", "HCO3", "O2", "H", "H2O", "ALG"),
                                        subst.norm = "ALG",
                                        nu.norm    = -1.0)
nu.death.ALG.complex <- calc.stoich.coef(alpha       = alpha.complex,
                                         name        = "death.ALG",
                                         subst       = c("NH4", "HPO4", "HCO3", "O2", "H", "H2O", "ALG", "POM"),
                                         subst.norm  = "ALG",
                                         nu.norm     = -1.0,
                                         constraints = list(c("ALG" = param$Y.ALG.death,
                                                              "POM" = 1.0)))
nu.growth.ZOO.complex <- calc.stoich.coef(alpha       = alpha.complex,
                                         name        = "growth.ZOO",
                                         subst       = c("NH4", "HPO4", "HCO3", "O2", "H", "H2O", "ALG", "ZOO", "POM"),
                                         subst.norm  = "ZOO",
                                         nu.norm     = 1.0,
                                         constraints = list(c("ZOO" = 1.0,
                                                              "ALG" = param$Y.ZOO),
                                                            c("POM" = 1.0,
                                                              "ALG" = param$f.e)))
nu.resp.ZOO.complex <- calc.stoich.coef(alpha       = alpha.complex,
                                        name        = "resp.ZOO",
                                        subst       = c("NH4", "HPO4", "HCO3", "O2", "H", "H2O", "ZOO"),
                                        subst.norm  = "ZOO",
                                        nu.norm     = -1.0)
nu.death.ZOO.complex <- calc.stoich.coef(alpha       = alpha.complex,
                                         name        = "death.ZOO",
                                         subst       = c("NH4", "HPO4", "HCO3", "O2", "H", "H2O", "ZOO", "POM"),
                                         subst.norm  = "ZOO",
                                         nu.norm     = -1.0,
                                         constraints = list(c("ZOO" = param$Y.ZOO.death,
                                                              "POM" = 1.0)))
nu.complex <- rbind(nu.growth.ALG.NH4.complex, 
                    nu.growth.ALG.NO3.complex,
                    nu.resp.ALG.complex,
                    nu.death.ALG.complex,
                    nu.growth.ZOO.complex,
                    nu.resp.ZOO.complex,
                    nu.death.ZOO.complex)
print(round(nu.complex, 3))




# --- introducing sulfur through sulfate --- #

param$alpha.S.ALG = 0.003
param$alpha.S.ZOO = 0.003
param$alpha.S.POM = 0.003

# update carbon factor such that factors sum to 1
param$alpha.C.ALG = 1 - (
    param$alpha.O.ALG
    + param$alpha.H.ALG
    + param$alpha.N.ALG
    + param$alpha.P.ALG
    + param$alpha.S.ALG
)
param$alpha.C.ZOO = 1 - (
    param$alpha.O.ZOO
    + param$alpha.H.ZOO
    + param$alpha.N.ZOO
    + param$alpha.P.ZOO
    + param$alpha.S.ZOO
)
param$alpha.C.POM = 1 - (
    param$alpha.O.POM
    + param$alpha.H.POM
    + param$alpha.N.POM
    + param$alpha.P.POM
    + param$alpha.S.POM
)

SO4 <- c(S      = 1,
         O      = 4,
         charge = -2)

ALG <- c(O = param$alpha.O.ALG / 16,
         H = param$alpha.H.ALG / 1,
         N = param$alpha.N.ALG / 14,
         P = param$alpha.P.ALG / 31,
         C = param$alpha.C.ALG / 12,
         S = param$alpha.S.ALG / 32)
ZOO <- c(O = param$alpha.O.ZOO / 16,
         H = param$alpha.H.ZOO / 1,
         N = param$alpha.N.ZOO / 14,
         P = param$alpha.P.ZOO / 31,
         C = param$alpha.C.ZOO / 12,
         S = param$alpha.S.ZOO / 32)
POM <- c(O = param$alpha.O.POM / 16,
         H = param$alpha.H.POM / 1,
         N = param$alpha.N.POM / 14,
         P = param$alpha.P.POM / 31,
         C = param$alpha.C.POM / 12,
         S = param$alpha.S.POM / 32)

subst.comp <- list(NH4  = NH4,
                   NO3  = NO3,
                   HPO4 = HPO4,
                   SO4  = SO4,
                   HCO3 = HCO3,
                   O2   = O2,
                   H    = H,
                   H2O  = H2O,
                   ALG  = ALG,
                   ZOO  = ZOO,
                   POM  = POM)

# build composition matrix
alpha.complex.SO4 <- calc.comp.matrix(subst.comp)
print(alpha.complex.SO4)

# update additional params
param$Y.ALG.death = min(1,
                        param$alpha.N.ALG / param$alpha.N.POM,
                        param$alpha.P.ALG / param$alpha.P.POM,
                        param$alpha.S.ALG / param$alpha.S.POM)
param$Y.ZOO.death = min(1,
                        param$alpha.N.ZOO / param$alpha.N.POM,
                        param$alpha.P.ZOO / param$alpha.P.POM,
                        param$alpha.S.ZOO / param$alpha.S.POM)

# calculate stoichiometric coefficients
nu.growth.ALG.NH4.complex.SO4 <- calc.stoich.coef(alpha      = alpha.complex.SO4,
                                                  name       = "growth.ALG.NH4",
                                                  subst      = c("NH4", "HPO4", "SO4", "HCO3", "O2", "H", "H2O", "ALG"),
                                                  subst.norm = "ALG",
                                                  nu.norm    = 1.0)
nu.growth.ALG.NO3.complex.SO4 <- calc.stoich.coef(alpha      = alpha.complex.SO4,
                                                  name       = "growth.ALG.NO3",
                                                  subst      = c("NO3", "HPO4", "SO4", "HCO3", "O2", "H", "H2O", "ALG"),
                                                  subst.norm = "ALG",
                                                  nu.norm    = 1.0)
nu.resp.ALG.complex.SO4 <- calc.stoich.coef(alpha      = alpha.complex.SO4,
                                            name       = "resp.ALG",
                                            subst      = c("NH4", "HPO4", "SO4", "HCO3", "O2", "H", "H2O", "ALG"),
                                            subst.norm = "ALG",
                                            nu.norm    = -1.0)
nu.death.ALG.complex.SO4 <- calc.stoich.coef(alpha       = alpha.complex.SO4,
                                             name        = "death.ALG",
                                             subst       = c("NH4", "HPO4", "SO4", "HCO3", "O2", "H", "H2O", "ALG", "POM"),
                                             subst.norm  = "ALG",
                                             nu.norm     = -1.0,
                                             constraints = list(c("ALG" = param$Y.ALG.death,
                                                                  "POM" = 1.0)))
nu.growth.ZOO.complex.SO4 <- calc.stoich.coef(alpha       = alpha.complex.SO4,
                                              name        = "growth.ZOO",
                                              subst       = c("NH4", "HPO4", "SO4", "HCO3", "O2", "H", "H2O", "ALG", "ZOO", "POM"),
                                              subst.norm  = "ZOO",
                                              nu.norm     = 1.0,
                                              constraints = list(c("ZOO" = 1.0,
                                                              "ALG" = param$Y.ZOO),
                                                            c("POM" = 1.0,
                                                              "ALG" = param$f.e)))
nu.resp.ZOO.complex.SO4 <- calc.stoich.coef(alpha       = alpha.complex.SO4,
                                            name        = "resp.ZOO",
                                            subst       = c("NH4", "HPO4", "SO4", "HCO3", "O2", "H", "H2O", "ZOO"),
                                            subst.norm  = "ZOO",
                                            nu.norm     = -1.0)
nu.death.ZOO.complex.SO4 <- calc.stoich.coef(alpha       = alpha.complex.SO4,
                                             name        = "death.ZOO",
                                             subst       = c("NH4", "HPO4", "SO4", "HCO3", "O2", "H", "H2O", "ZOO", "POM"),
                                             subst.norm  = "ZOO",
                                             nu.norm     = -1.0,
                                             constraints = list(c("ZOO" = param$Y.ZOO.death,
                                                                  "POM" = 1.0)))
nu.complex.SO4 <- rbind(nu.growth.ALG.NH4.complex.SO4, 
                        nu.growth.ALG.NO3.complex.SO4,
                        nu.resp.ALG.complex.SO4,
                        nu.death.ALG.complex.SO4,
                        nu.growth.ZOO.complex.SO4,
                        nu.resp.ZOO.complex.SO4,
                        nu.death.ZOO.complex.SO4)
print(signif(nu.complex.SO4, 3))

File: stoichiometry.py
import numpy as np
import pandas as pd
from helpers.convert import mass_to_molar, molar_to_mass


# --- simple stoichiometric solution --- #

compositions = {
    'HPO4': {
        'P': 1.0,   # gP/gHPO4-P
        'N': 0.0    # gN/gHPO4-P
    },
    'NH4': {
        'P': 0.0,   # gP/gN4-N
        'N': 1.0    # gN/gN4-N
    },
    'ALG': {
        'P': 0.01,  # gP/gALG
        'N': 0.1    # gN/gALG
    }
}

substances = list(compositions.keys())
elemental_constituents = list({                            
    e for comp in compositions.values()
    for e in comp
})


# define composition matrix
alpha = np.array([
    [compositions[s].get(e, 0.0) for s in substances]
    for e in elemental_constituents
])

alpha_df = pd.DataFrame(
    alpha,
    index= elemental_constituents,
    columns= substances
).round(3)


# fix a mass factor for normalization and solve for process stoichiometric coefficients
normalizer = np.zeros(len(substances))
normalizer[-1] = 1.0

A = np.vstack([alpha, normalizer])
b = normalizer

nu_growth = np.linalg.solve(A, b)
nu_resp = -nu_growth

# define stoichiometric coefficients matrix
nu = np.vstack([
    nu_growth,
    nu_resp
])

nu_df = pd.DataFrame(
    nu,
    index= ['growth of algae', 'respiration of algae'],
    columns= substances
).round(3)




# --- complex stoichiometric solution --- #

params = {
    'Y_ZOO': 0.2,  # gZOO/gALG
    'f_e': 0.4     # gPOM/gALG
}

elemental_constituents = ['H', 'N', 'O', 'P', 'C']
constraints = elemental_constituents + ['charge']

organism_compositions = {
    'ALG': {
        'O': 0.50,  # gO/gALG
        'H': 0.07,  # gH/gALG
        'N': 0.06,  # gN/gALG
        'P': 0.005  # gP/gALG
    },
    'ZOO': {
        'O': 0.50,  # gO/gZOO
        'H': 0.07,  # gH/gZOO
        'N': 0.06,  # gN/gZOO
        'P': 0.01   # gP/gZOO
    },
    'POM': {
        'O': 0.40,  # gO/gPOM
        'H': 0.07,  # gH/gPOM
        'N': 0.04,  # gN/gPOM
        'P': 0.007  # gP/gPOM
    }
}

organisms = list(organism_compositions.keys())

# choose carbon factors to guarantee factors sum to unity
for org in organisms:
    organism_compositions[org]['C'] = 1 - sum([
        organism_compositions[org][e]
        for e in elemental_constituents if e != 'C'
    ])

substance_compositions = {
    'NH4': {
        'N': 1,
        'H': 4,
        'charge': 1
    },
    'NO3': {
        'N': 1,
        'O': 3,
        'charge': -1
    },
    'HPO4': {
        'H': 1,
        'P': 1,
        'O': 4,
        'charge': -2
    },
    'HCO3': {
        'H': 1,
        'C': 1,
        'O': 3,
        'charge': -1
    },
    'O2': {
        'O': 2
    },
    'H': {
        'H': 1,
        'charge': 1
    },
    'H2O': {
        'H': 2,
        'O': 1
    }
}

substances = list(substance_compositions.keys())

# convert organisms from mass to molar
for org in organisms:
    organism_compositions[org] = {
        e: mass_to_molar(e, val)
        for e, val in organism_compositions[org].items()
    }

# compositions of substances and organisms combined
compositions = {**organism_compositions, **substance_compositions}
substances += organisms


# define composition matrix
alpha = np.array([
    [compositions[s].get(c, 0.0) for s in substances]
    for c in constraints
])

alpha_df = pd.DataFrame(
    alpha,
    index= constraints,
    columns= substances
).round(3)


def get_coeffs_df(process, subs_and_orgs, normalized_substance, norm, additional_constraints= None):
    alpha = np.array([
        [compositions[s].get(c, 0.0) for s in subs_and_orgs]
        for c in constraints
    ])

    normalization_constraint = np.zeros(len(subs_and_orgs))
    normalization_constraint[subs_and_orgs.index(normalized_substance)] = norm

    rows = [alpha, normalization_constraint]
    if additional_constraints is not None:
        rows.extend(additional_constraints)

    A = np.vstack(rows)
    b = np.zeros(A.shape[0])
    b[len(constraints)] = 1.0

    coeffs = np.linalg.solve(A, b)

    return pd.DataFrame(
        [coeffs],
        index= [process],
        columns= subs_and_orgs
    )


# - get stoichiometric coefficients for each process -
nu_df_rows = []

# algae growth with ammonium as nitrogen source
ALG_growth_NH4_substances = [s for s in substances if s not in {'NO3', 'ZOO', 'POM'}]
nu_df_rows.append(get_coeffs_df(
    'algae growth (NH4)',
    ALG_growth_NH4_substances,
    'ALG', 1.0
))

# algae growth with nitrate as nitrogen source
ALG_growth_NO3_substances = [s for s in substances if s not in {'NH4', 'ZOO', 'POM'}]
nu_df_rows.append(get_coeffs_df(
    'algae growth (NO3)',
    ALG_growth_NO3_substances,
    'ALG', 1.0
))

# algae respiration with ammonium as nitrogen source
nu_df_rows.append(get_coeffs_df(
    'algae respiration (NH4)',
    ALG_growth_NH4_substances,
    'ALG', -1.0
))

# algae respiration with nitrate as nitrogen source
nu_df_rows.append(get_coeffs_df(
    'algae respiration (NO3)',
    ALG_growth_NO3_substances,
    'ALG', -1.0
))

# algae death
ALG_death_substances = [s for s in substances if s not in {'NO3', 'ZOO'}]

Y_ALG_death = min(1, 
                  compositions['ALG']['N'] / compositions['POM']['N'],
                  compositions['ALG']['P'] / compositions['POM']['P'])

ALG_death_constraint = np.zeros(len(ALG_death_substances))
ALG_death_constraint[ALG_death_substances.index('POM')] = 1.0
ALG_death_constraint[ALG_death_substances.index('ALG')] = Y_ALG_death

nu_df_rows.append(get_coeffs_df(
    'algae death',
    ALG_death_substances,
    'ALG', -1.0,
    additional_constraints= [ALG_death_constraint]
))

# zooplankton growth with ammonium as nitrogen source
ZOO_growth_substances = [s for s in substances if s != 'NO3']

ZOO_growth_constraint = np.zeros(len(ZOO_growth_substances))
ZOO_growth_constraint[ZOO_growth_substances.index('ZOO')] = 1.0
ZOO_growth_constraint[ZOO_growth_substances.index('ALG')] = params['Y_ZOO']

POM_constraint = np.zeros(len(ZOO_growth_substances))
POM_constraint[ZOO_growth_substances.index('POM')] = 1.0
POM_constraint[ZOO_growth_substances.index('ALG')] = params['f_e']

nu_df_rows.append(get_coeffs_df(
    'zooplankton growth',
    ZOO_growth_substances,
    'ZOO', 1.0,
    additional_constraints= [ZOO_growth_constraint, POM_constraint]
))

# zooplankton respiration
ZOO_respiration_substances = [s for s in substances if s not in {'NO3', 'ALG', 'POM'}]
nu_df_rows.append(get_coeffs_df(
    'zooplankton respiration',
    ZOO_respiration_substances,
    'ZOO', -1.0
))

# zooplankton death
ZOO_death_substances = [s for s in substances if s not in {'NO3', 'ALG'}]

Y_ZOO_death = min(1, 
                  compositions['ZOO']['N'] / compositions['POM']['N'],
                  compositions['ZOO']['P'] / compositions['POM']['P'])

ZOO_death_constraint = np.zeros(len(ZOO_death_substances))
ZOO_death_constraint[ZOO_death_substances.index('POM')] = 1.0
ZOO_death_constraint[ZOO_death_substances.index('ZOO')] = Y_ZOO_death

nu_df_rows.append(get_coeffs_df(
    'zooplankton death',
    ZOO_death_substances,
    'ZOO', -1.0,
    additional_constraints= [ZOO_death_constraint]
))

nu_df = pd.concat(nu_df_rows, axis= 0).fillna(0).round(3).reindex(columns= substances)
print(nu_df)




# --- introducing sulfur through sulfate --- #

elemental_constituents += ['S']
constraints = elemental_constituents + ['charge']

# convert organisms back to mass
for org in organisms:
    organism_compositions[org] = {
        e: molar_to_mass(e, val)
        for e, val in organism_compositions[org].items()
    }


organism_compositions['ALG']['S'] = 0.003
organism_compositions['ZOO']['S'] = 0.003
organism_compositions['POM']['S'] = 0.003

# update carbon factors to guarantee factors sum to unity
for org in organisms:
    organism_compositions[org]['C'] = 1 - sum([
        organism_compositions[org][e]
        for e in elemental_constituents if e != 'C'
    ])

substance_compositions['SO4'] = {
    'S': 1,
    'O': 4,
    'charge': -2
}

# update substances list
substances = list(substance_compositions.keys())

# convert organisms from mass to molar
for org in organisms:
    organism_compositions[org] = {
        e: mass_to_molar(e, val)
        for e, val in organism_compositions[org].items()
    }

# compositions of substances and organisms combined
compositions = {**organism_compositions, **substance_compositions}
substances += organisms


# define composition matrix
alpha = np.array([
    [compositions[s].get(c, 0.0) for s in substances]
    for c in constraints
])

alpha_df = pd.DataFrame(
    alpha,
    index= constraints,
    columns= substances
).round(5)


# - get stoichiometric coefficients for each process -
nu_df_rows = []

# algae growth with ammonium as nitrogen source
ALG_growth_NH4_substances = [s for s in substances if s not in {'NO3', 'ZOO', 'POM'}]
nu_df_rows.append(get_coeffs_df(
    'algae growth (NH4)',
    ALG_growth_NH4_substances,
    'ALG', 1.0
))

# algae growth with nitrate as nitrogen source
ALG_growth_NO3_substances = [s for s in substances if s not in {'NH4', 'ZOO', 'POM'}]
nu_df_rows.append(get_coeffs_df(
    'algae growth (NO3)',
    ALG_growth_NO3_substances,
    'ALG', 1.0
))

# algae respiration with ammonium as nitrogen source
nu_df_rows.append(get_coeffs_df(
    'algae respiration (NH4)',
    ALG_growth_NH4_substances,
    'ALG', -1.0
))

# algae respiration with nitrate as nitrogen source
nu_df_rows.append(get_coeffs_df(
    'algae respiration (NO3)',
    ALG_growth_NO3_substances,
    'ALG', -1.0
))

# algae death
ALG_death_substances = [s for s in substances if s not in {'NO3', 'ZOO'}]

Y_ALG_death = min(1, 
                  compositions['ALG']['N'] / compositions['POM']['N'],
                  compositions['ALG']['P'] / compositions['POM']['P'],
                  compositions['ALG']['S'] / compositions['POM']['S'])

ALG_death_constraint = np.zeros(len(ALG_death_substances))
ALG_death_constraint[ALG_death_substances.index('POM')] = 1.0
ALG_death_constraint[ALG_death_substances.index('ALG')] = Y_ALG_death

nu_df_rows.append(get_coeffs_df(
    'algae death',
    ALG_death_substances,
    'ALG', -1.0,
    additional_constraints= [ALG_death_constraint]
))

# zooplankton growth with ammonium as nitrogen source
ZOO_growth_substances = [s for s in substances if s != 'NO3']

ZOO_growth_constraint = np.zeros(len(ZOO_growth_substances))
ZOO_growth_constraint[ZOO_growth_substances.index('ZOO')] = 1.0
ZOO_growth_constraint[ZOO_growth_substances.index('ALG')] = params['Y_ZOO']

POM_constraint = np.zeros(len(ZOO_growth_substances))
POM_constraint[ZOO_growth_substances.index('POM')] = 1.0
POM_constraint[ZOO_growth_substances.index('ALG')] = params['f_e']

nu_df_rows.append(get_coeffs_df(
    'zooplankton growth',
    ZOO_growth_substances,
    'ZOO', 1.0,
    additional_constraints= [ZOO_growth_constraint, POM_constraint]
))

# zooplankton respiration
ZOO_respiration_substances = [s for s in substances if s not in {'NO3', 'ALG', 'POM'}]
nu_df_rows.append(get_coeffs_df(
    'zooplankton respiration',
    ZOO_respiration_substances,
    'ZOO', -1.0
))

# zooplankton death
ZOO_death_substances = [s for s in substances if s not in {'NO3', 'ALG'}]

Y_ZOO_death = min(1, 
                  compositions['ZOO']['N'] / compositions['POM']['N'],
                  compositions['ZOO']['P'] / compositions['POM']['P'],
                  compositions['ZOO']['S'] / compositions['POM']['S'])

ZOO_death_constraint = np.zeros(len(ZOO_death_substances))
ZOO_death_constraint[ZOO_death_substances.index('POM')] = 1.0
ZOO_death_constraint[ZOO_death_substances.index('ZOO')] = Y_ZOO_death

nu_df_rows.append(get_coeffs_df(
    'zooplankton death',
    ZOO_death_substances,
    'ZOO', -1.0,
    additional_constraints= [ZOO_death_constraint]
))

nu_df = pd.concat(nu_df_rows, axis= 0).fillna(0).round(5).reindex(columns= substances)
print(nu_df)
